lib/src/Uncertainty/Algorithm/Sensitivity/HSICEstimatorImplementation.cxx

```diff
@@ -0,0 +1,472 @@
+{
+ // Nothing
+}
+
+
+/* Default constructor */
+HSICEstimatorImplementation::HSICEstimatorImplementation(
+  const CovarianceModelCollection & covarianceList
+  , const Sample & X
+  , const Sample & Y
+  , const HSICStat & estimatorType, const Function & weightFunction )
+  : PersistentObject()
+  , covarianceList_(covarianceList)
+  , inputSample_(X)
+  , outputSample_(Y)
+  , estimatorType_(estimatorType)
+  , weightFunction_(weightFunction)
+  , n_(X.getSize())
+  , inputDimension_(X.getDimension())
+  , HSIC_XY_()
+  , HSIC_XX_()
+  , HSIC_YY_()
+  , R2HSICIndices_()
+  , PValuesPermutation_()
+  , permutationSize_(ResourceMap::GetAsUnsignedInteger("HSICEstimatorImplementation-PermutationSize"))
+  , isAlreadyComputedIndices_(false)
+  , isAlreadyComputedPValuesPermutation_(false)
+{
+  if(covarianceList_.getSize() != (inputSample_.getDimension() + outputSample_.getDimension())) throw InvalidDimensionException(HERE) << "The number of covariance momdels is the dimension of the input +1";
+  if(outputSample_.getDimension() != 1) throw InvalidDimensionException(HERE) << "The dimension of the output is 1.";
+  if(inputSample_.getSize() != outputSample_.getSize()) throw InvalidDimensionException(HERE) << "Input and output samples must have the same size";
+
+}
+
+/* Default constructor */
+HSICEstimatorImplementation::HSICEstimatorImplementation(
+  const CovarianceModelCollection & covarianceList
+, const Sample & X
+, const Sample & Y
+, const HSICStat & estimatorType)
+  : PersistentObject()
+  , covarianceList_(covarianceList)
+  , inputSample_(X)
+  , outputSample_(Y)
+  , estimatorType_(estimatorType)
+  , n_(X.getSize())
+  , inputDimension_(X.getDimension())
+  , HSIC_XY_ ()
+  , HSIC_XX_ ()
+  , HSIC_YY_ ()
+  , R2HSICIndices_ ()
+  , permutationSize_(ResourceMap::GetAsUnsignedInteger("HSICEstimatorImplementation-PermutationSize"))
+  , isAlreadyComputedIndices_(false)
+  , isAlreadyComputedPValuesPermutation_(false)
+{
+  if(covarianceList_.getSize() != (inputSample_.getDimension() + outputSample_.getDimension())) throw InvalidDimensionException(HERE) << "The number of covariance momdels is the dimension of the input +1";
+  if(outputSample_.getDimension() != 1) throw InvalidDimensionException(HERE) << "The dimension of the output is 1.";
+  if(inputSample_.getSize() != outputSample_.getSize()) throw InvalidDimensionException(HERE) << "Input and output samples must have the same size";
+
+  weightFunction_  = SymbolicFunction("x", "1");
+}
+
+/* Virtual constructor */
+HSICEstimatorImplementation * HSICEstimatorImplementation::clone() const
+{
+  return new HSICEstimatorImplementation(*this);
+}
+
+/* Compute the weight matrix from the weight function */
+SquareMatrix HSICEstimatorImplementation::computeWeightMatrix(const Sample & Y) const
+{
+  const Sample wY(weightFunction_(Y));
+  const Scalar meanWY(wY.computeMean()[0]);
+  SquareMatrix mat(n_);
+  mat.setDiagonal(wY.asPoint() / meanWY);
+  return mat;
+}
```

Comment:
Following our last CT meeting:

- This implementation of the method and the `weightFunction_` attribute should move to `HSICEstimatorConditionalSensitivity`.
- `HSICEstimatorImplementation::computeWeightMatrix` should simply throw a `NotYetImplemented` exception.

Is that correct?

Reply:
Please note that both `HSICEstimatorTargetSensitivity` and `HSICEstimatorGlobalSensitivity` require either a working`computeWeightMatrix` method or, at least, a sort of `weightMatrix_` attribute in order to call the `computeHSICIndex` method of the `HSICStat` class.

Reply:
@JPelamatti There is no need for a `weightMatrix_` attribute, but we need 4 `computeWeightMatrix` methods:

- in `HSICEstimatorImplementation`: will throw a `NotYetImplementedException`
- in `HSICEstimatorConditionalSensitivity`: will use the implementation above, the one which relies on a `weightFunction_` atttribute (declared as an `HSICEstimatorConditionalSensitivity` attribute rather than an `HSICEstimatorImplementation` attribute)
- in `HSICEstimatorGlobalSensitivity`: already implemented - nothing to change
- in `HSICEstimatorTargetSensitivity`: already implemented - nothing to change

That is, I believe, the plan proposed by @regislebrun and which we agreed to.

Reply:
@josephmure I apologize, I misunderstood your message and thought you wanted to remove the impementations in `HSICEstimatorGlobalSensitivity` and `HSICEstimatorTargetSensitivity`.
In this case, I totally agree.